dom.bs

```diff
@@ -2752,6 +2812,16 @@ before a <var>child</var>, with an optional <i>suppress observers flag</i>, run
  <var>parent</var> with <var>nodes</var>, « », <var>previousSibling</var>, and <var>child</var>.
 
  <li><p>Run the <a>children changed steps</a> for <var>parent</var>.
+
+ <li>
+  <p>For each <var>node</var> in <var>nodes</var>, in <a>tree order</a>:
+
+  <ol>
+   <li><p>For each <a>shadow-including inclusive descendant</a> <var>inclusiveDescendant</var> of
+   <var>node</var>, in <a>shadow-including tree order</a>, run the <a>post-insertion steps</a> with
+   <var>inclusiveDescendant</var>.
+  </ol>
+ </li>
```

Comment:
> Is this the correct order relative to mutation observers? This should be observable using `takeRecords()`.

Good catch. Yes, Chrome and Firefox both can observe the _queued_ record this way. I am not actually sure the WPTs assert this, so let me modify a few to include this check.

> I also still worry about this not being entirely consistent with the removal case.

Hmm, could you elaborate on this concern? I don't know what is inconsistent about it. Assuming we have no opportunities to execute script during removal, we shouldn't need to move anything to an equivalent post-removal steps. And post-removal steps seem risky anyways, since we'd be operating exclusively on detached elements. The true equivalent for removal would be to let each node run their own script-observable non-script-executing side-effects one-by-one before any of them get removed, but I feel like as long as none of them are executing script, then that shouldn't actually be observable? And we certainly don't want to run script for each element just immediately before its removal, since that's the worst part of mutation events.

Reply:
Currently only chromium fires `blur` events when removing, and we're checking if that's safe to fix. The only other existing script side effect is `pagehide` in same-process iframes. 

Both are not in the spec - if just reading the standard and not implementation, there are 0 script side effects to removing nodes (unless we missed a spot).

Reply:
Yeah, please see [this part of the OP](https://github.com/whatwg/dom/pull/1261#:~:text=Dealing%20with%20%22removal%22%20steps) of this PR that discusses the current state of removal. I promised we wouldn't proceed with insertion without taking a reasonable look at the removal story!

Reply:
Gecko triggers script execution after all the insertions have been done. 
But there should be just one MutationRecord ,  that should be covered by DOM spec when dealing with adding DocumentFragments and such. MutationRecord has all the inserted and removed nodes in separate lists.

Reply:
Hmm, this model isn't exactly what Gecko has. We use script runners to run stuff ,which might run scripts, at safe time. I wonder if that model would be easier here.

Reply:
I don't know what a script runner is. But if you have a concrete test case that shows how Gecko deviates from the spec here, that would be great. Two points though:

 - This PR doesn't actually make changes to script execution. That would be https://github.com/whatwg/html/pull/10188, which I'd be grateful for your review on as well!
 - The combined behavior change of both this PR and https://github.com/whatwg/html/pull/10188 is captured by [these tests](https://wpt.fyi/results/dom/nodes/insertion-removing-steps?label=experimental&label=master&aligned). We'd love for you to weigh on the test expectations over in https://github.com/whatwg/html/pull/10188, to see if any of the significant Firefox failures are a result of inappropriate test expectations.

Reply:
Script runners are like the nano tasks which were discussed years ago. Basically a thing which runs something which may trigger script execution, and runs at safe time. Very similar to post-insertion steps. But the key difference is that those aren't bound strictly to any node.  And you may create script runners (and often do) during insertion steps.

Reply:
Script runners are just like tasks which are run when there is first chance to run scripts.  The nice thing there is that there can be many of them, and they don't need to be bound to any node. It is unclear to me how "post insertion steps" would work if there are multiple things which need to be done at that time.

Reply:
Given that we don't often add new post-insertion steps/script runners, and the fact that there aren't that many of those today, is there an observable difference between that and post-insertion steps? To me it doesn't seem material whether we spec it this way or that way, it's a way to express that iframes & scripts never run scriptable side-effects in the middle of the DOM update.